src/array_like.rs

```diff
@@ -0,0 +1,111 @@
+//! TODO
+
+use crate::{
+    ndarray::{Array, Array1, ArrayView, Axis, Dimension},
+    Element, PyReadonlyArray,
+};
+use ndarray::{Ix1, Ix2, Ix3, Ix4, Ix5, Ix6, IxDyn};
+use pyo3::{exceptions::PyValueError, FromPyObject, PyAny, PyResult};
+
+/// TODO
+#[derive(Debug)]
+pub struct PyArrayLike<'py, T, D>(ArrayLike<'py, T, D>)
+where
+    T: Element,
+    D: Dimension;
+
+#[derive(Debug)]
+enum ArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    Reference(PyReadonlyArray<'py, T, D>),
+    Converted(Array<T, D>),
+}
+
+impl<'py, T, D> PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+{
+    /// TODO
+    pub fn into_owned_array(self) -> Array<T, D> {
+        match self.0 {
+            ArrayLike::Converted(x) => x,
+            ArrayLike::Reference(x) => x.to_owned_array(),
+        }
+    }
+
+    /// TODO
+    pub fn as_view(&self) -> ArrayView<T, D> {
+        match &self.0 {
+            ArrayLike::Converted(x) => x.view(),
+            ArrayLike::Reference(x) => x.as_array(),
+        }
+    }
+
+    /// TODO
+    pub fn is_reference(&self) -> bool {
+        match &self.0 {
+            ArrayLike::Reference(_) => true,
+            ArrayLike::Converted(_) => false,
+        }
+    }
+}
+
+impl<'py, T, D> FromPyObject<'py> for PyArrayLike<'py, T, D>
+where
+    T: Element,
+    D: Dimension,
+    Vec<T>: FromPyObject<'py>,
+{
+    fn extract(ob: &'py PyAny) -> PyResult<Self> {
+        if let Ok(array) = ob.extract::<PyReadonlyArray<T, D>>() {
+            return Ok(PyArrayLike(ArrayLike::Reference(array)));
+        }
+
+        if D::NDIM.unwrap_or(1) == 1 {
+            if let Ok(array) = ob.extract::<Vec<T>>() {
+                let res = Array1::from(array)
+                    .into_dimensionality()
+                    .map_err(|_| PyValueError::new_err("Dimension incompatible"))?;
+                return Ok(PyArrayLike(ArrayLike::Converted(res)));
+            }
+        }
+
+        if D::NDIM == Some(1) {
+            return Err(PyValueError::new_err("Expected one dimensional array"));
+        }
+
+        if let Ok(py_list) = ob.extract::<Vec<&PyAny>>() {
+            let sub_arrays = py_list
+                .into_iter()
+                .map(|x| x.extract::<PyArrayLike<T, D::Smaller>>())
+                .collect::<Result<Vec<_>, _>>()?;
+            let sub_array_views = sub_arrays.iter().map(|x| x.as_view()).collect::<Vec<_>>();
```

Comment:
While this is clever, it potentially does quite a lot of temporary allocations. I wonder whether calling the Python function `asarray` would be a better fallback here then? There is some overhead to calling back into Python, but we automatically benefit from upstream changes. The really important list/iterator fast paths could still avoid this overhead.

Reply:
I did not know this was a possibility at all, but I definitely agree! I am quite sure that the implementation of `asarray` should be way more efficient than mine. I'm preparing a followup commit..